src/webserver/routing.rs

```diff
@@ -0,0 +1,496 @@
+use awc::http::Uri;
+use std::ffi::OsStr;
+use std::path::PathBuf;
+use RoutingAction::{Error, Execute, NotFound, Redirect, Serve};
+
+const INDEX: &'static str = "index.sql";
+const NOT_FOUND: &'static str = "404.sql";
+const EXECUTION_EXTENSION: &'static str = "sql";
+const FORWARD_SLASH: &'static str = "/";
+
+#[derive(Debug, PartialEq)]
+pub enum RoutingAction {
+    Error(String),
+    Execute(PathBuf),
+    NotFound(PathBuf),
+    Redirect(Uri),
+    Serve(PathBuf),
+}
+
+pub trait ExecutionStore {
+    async fn contains(&self, path: &PathBuf) -> bool;
+}
+
+pub trait RoutingConfig {
+    fn prefix(&self) -> &Uri;
+}
+
+pub async fn calculate_route<T, C>(uri: Uri, store: &T, config: &C) -> RoutingAction
```

Comment:
why are we taking a Uri ? We should not have to build a Uri to make the routing, we should be able to operate only on the path.

Reply:
My thinking here was:
1/ trying to increase the guarantee on the input; specifically I don't want to deal with data that isn't a valid uri (encoding issues and what-not).
2/ the contract is uri in to uri out (redirects) or pathbuf out (serve files) - so it felt appropriate
3/ `Uri` makes it easier to use path-only for files and path+query+fragment for redirects (was trying to avoid string splitting and so-forth)

`Uri` itself isn't the most usable or friendly type (it lacks basic stuff like append), but it felt a little better than a plain `&str`.

Reply:
I think str or bytes are better than building a fake uri just for calling the function. We should not need anything more than a path to make our routing decision.

Reply:
ok I can change it; I wasn't particularly wedded to it. 

I had assumed we'd just pass in the value from `actix_web::service::ServiceRequest::uri()` - it seems odd to downcast it to a `&str`?

Reply:
This uri is reconstructed, and not particularly meaningful (in particular, it is not the uri displayed in the browser's address bar if the service is deployed behind a reverse proxy. I don't think using it is helpful. We won't downcast it, we'll just path the incoming request path unmodified.

Reply:
Oh, ok I think I've misunderstood the redirect requirement. I had assumed that we would maintain the query and fragment when redirecting ([as per these tests](https://github.com/guspower/SQLpage/blob/85b4dabf759e1771434118ef520581dc50984e28/src/webserver/routing.rs#L374)). If we don't need this then yep we can just use the `path`.

Reply:
No, no, you are right ! We should maintain them. We can take a [`PathAndQuery`](https://docs.rs/actix-web/latest/actix_web/http/uri/struct.PathAndQuery.html) as input. This way we do not pass a fake host and scheme, but we don't have to parse a string either.

Reply:
Right cool, I'll have a go with `PathAndQuery` this afternoon. Onwards!